### PR TITLE
Pr batch env

### PR DESCRIPTION
**Pull request by runjerry**
GYM to Pytorch wrappers and parallel environments handling for ALF Pytorch.

### CODE REVIEW

File: alf/environments/parallel_torch_environment.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            env = ProcessEnvironment(ctor, flatten=flatten)
+            env_id = torch.tensor(env_id, dtype=torch.int32)
+            env._env_id = env_id
+            self._envs.append(env)
+            self._env_ids.append(env_id)
+        self._num_envs = len(env_constructors)
+        self._blocking = blocking
+        self._start_serially = start_serially
+        self.start()
+        self._action_spec = self._envs[0].action_spec()
```
Comment (le-horizon) [inline alf/environments/parallel_torch_environment.py]: just wondering, isn't it strange that the specs are not batched tensor specs, but instead the specs of the individual env?

File: alf/environments/suite_dmlab_test.py [hunk 2]
```diff
@@ -60,39 +65,44 @@ def test_action_discretize(self, scene, action_config, action_length):
...
+                scene='lt_chasm',
+                gym_env_wrappers=[
+                    gym_wrappers.FrameGrayScale, gym_wrappers.FrameResize,
+                    gym_wrappers.FrameStack
+                ],
+                wrap_with_process=False)
+
+        self._env = parallel_torch_environment.ParallelTorchEnvironment(
+            [ctor] * 2)
+        # env = tf_py_environment.TFPyEnvironment(self._env)
```
Comment (le-horizon) [inline alf/environments/suite_dmlab_test.py]: Add TODO to modify the test back?

File: alf/environments/suite_dmlab_test.py [hunk 3]
```diff
@@ -60,39 +65,44 @@ def test_action_discretize(self, scene, action_config, action_length):
...
+                scene=scene,
+                gym_env_wrappers=[gym_wrappers.FrameResize],
+                wrap_with_process=False)
 
-        self._env = parallel_py_environment.ParallelPyEnvironment([ctor] * 4)
-        env = tf_py_environment.TFPyEnvironment(self._env)
-        self.assertEqual((84, 84, 3), env.observation_spec().shape)
+        self._env = parallel_torch_environment.ParallelTorchEnvironment(
+            [ctor] * 4)
+        # env = tf_py_environment.TFPyEnvironment(self._env)
```
Comment (le-horizon) [inline alf/environments/suite_dmlab_test.py]: Same here?

File: alf/environments/suite_dmlab_test.py [hunk 4]
```diff
@@ -60,39 +65,44 @@ def test_action_discretize(self, scene, action_config, action_length):
...
+        # random_policy = random_tf_policy.RandomTFPolicy(
+        #     env.time_step_spec(), env.action_spec())
 
-        driver = dynamic_step_driver.DynamicStepDriver(
-            env=env, policy=random_policy, observers=None, num_steps=10)
+        # driver = dynamic_step_driver.DynamicStepDriver(
+        #     env=env, policy=random_policy, observers=None, num_steps=10)
 
-        driver.run(maximum_iterations=10)
+        # driver.run(maximum_iterations=10)
```
Comment (le-horizon) [inline alf/environments/suite_dmlab_test.py]: Add  a TODO to re-enable this block?

File: alf/environments/suite_gym_test.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        self.assertIsInstance(env, torch_wrappers.TimeLimit)
+        self.assertEqual(5, env._duration)
+
+    def test_load_atari(self):
+        env = suite_gym.load(
+            'BreakoutNoFrameskip-v4',
+            gym_env_wrappers=(DMAtariPreprocessing, FrameStack))
+        self.assertIsInstance(env, TorchEnvironment)
+
+    # # unittest does not have test_src_dir_path
```
Comment (le-horizon) [inline alf/environments/suite_gym_test.py]: add TODO to add the test back?

File: alf/environments/torch_wrappers.py [hunk 6]
```diff
@@ -0,0 +1,950 @@
...
+import cProfile
+import numpy as np
+import six
+import torch
+import gin
+
+from alf.tensor_specs import is_bounded
+from alf.data_structures import StepType, TimeStep
+from alf.environments import torch_environment
+import alf.nest as nest
```
Comment (emailweixu) [inline alf/environments/torch_wrappers.py]: only keep what have been used in ALF

File: alf/data_structures.py [hunk 7]
```diff
@@ -143,6 +145,174 @@ def timestep_last(observation, prev_action, reward, discount, env_id):
...
+    first_observation = nest.flatten(observation)[0]
+    assert torch.is_tensor(first_observation)
+    first_action = nest.flatten(action)[0]
+    assert torch.is_tensor(first_action)
+
+    # TODO(b/130245199): If reward.shape.rank == 2, and static
+    # batch sizes are available for both first_observation and reward,
+    # check that these match.
+    reward = torch.tensor(reward, dtype=torch.float32)
+    if reward.dim() is None or reward.dim() > 1:
```
[the remark this answers is not in the document]
Comment (emailweixu) [reply alf/data_structures.py]:
```python
is_valid_spec = lambda spec: isinstance(spec, (TensorSpec, BoundedTensorSpec))
assert all(map(is_valid_spec, alf.nest.flatten(observation_spec))), "...."
```

File: alf/tensor_specs.py [hunk 8]
```diff
@@ -17,15 +17,52 @@
...
+    Returns:
+      A nest of sampled values following the TensorSpec definition.
+    """
+
+    def sample_fn(spec):
+        shape = spec.shape
+        if outer_dims is not None:
+            shape = tuple(outer_dims) + shape
+        spec = BoundedTensorSpec.from_spec(spec)
+        spec = BoundedTensorSpec(shape, spec.dtype, spec.minimum, spec.maximum)
```
Comment (hnyu) [inline alf/tensor_specs.py]: Why do you want to re-create the `spec` here?

File: alf/tensor_specs.py [hunk 9]
```diff
@@ -17,15 +17,52 @@
...
+def is_bounded(spec):
+    return isinstance(spec, BoundedTensorSpec)
+
+
+def is_discrete(spec):
+    assert isinstance(spec, TensorSpec)
+    return spec.dtype.is_integer
+
+
+def is_continuous(spec):
```
Comment (hnyu) [inline alf/tensor_specs.py]: These three functions are already implemented as class members.

File: alf/tensor_specs.py [hunk 10]
```diff
@@ -218,10 +255,12 @@ def is_bounded(cls):
...
-        assert isinstance(spec, BoundedTensorSpec)
-        minimum = getattr(spec, "minimum")
-        maximum = getattr(spec, "maximum")
-        return BoundedTensorSpec(spec.shape, spec.dtype, minimum, maximum)
+        assert isinstance(spec, (TensorSpec, BoundedTensorSpec))
+        if hasattr(spec, 'minimum') and hasattr(spec, 'maximum'):
+            return BoundedTensorSpec(spec.shape, spec.dtype, spec.minimum,
+                                     spec.maximum)
+
+        return BoundedTensorSpec(spec.shape, spec.dtype)
```
Comment (hnyu) [inline alf/tensor_specs.py]: There are always default `min` and `max` values for `BoundedTensorSpec`. So you don't need `if`

**Comment by runjerry** [comment]
new PR opened [#423](https://github.com/HorizonRobotics/alf/pull/423)

**Pull request closed without merging**